src/pam_ibmacf.cpp

```diff
@@ -266,38 +261,49 @@ void setSerialNumberProperty(const string& obj){
     mSerialNumber = obj;
 }
 #else
-int readFieldModeProperty(const string& obj, const string& inf,
-                          const string& prop, pam_handle_t* pamh)
+int readFieldMode(pam_handle_t* pamh)
 {
-    bool propBool = false;
-    std::string object = obj;
-    auto bus = sdbusplus::bus::new_default();
-    auto properties = bus.new_method_call(
-        "xyz.openbmc_project.Software.BMC.Updater", object.c_str(),
-        "org.freedesktop.DBus.Properties", "Get");
-    properties.append(inf);
-    properties.append(prop);
-    try
+    FILE* pipe = popen("fw_printenv -n fieldmode 2>&1", "r");
+    if (!pipe)
     {
-        auto result = bus.call(properties);
-        if (!result.is_method_error())
+        pam_syslog(pamh, LOG_ERR, "popen failed\n");
+        return PAM_SYSTEM_ERR;
+    }
+    std::array<char, 512> buffer;
+    std::stringstream result;
+    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr)
+    {
+        result << buffer.data();
+    }
+    int stat = pclose(pipe);
+    if (WIFEXITED(stat))
```

Comment:
I've not seen these WIFEXITED and WEXITSTATUS maros before. Googling shows them to be kind of IBM specific. Why not just a check to make sure pclose returns 0 and call it good?

Reply:
The W* macros are Linux, defined by the wait() syscall, not IBM specific.
The pclose function technically returns a status, not the program return code (rc).
So the way way I coded it is correct to check the status first, because the return code (rc) will only be meaningful if the status is EXITED.

The code structure is to differentiate between two cases:
1. Line 291: The command exits normally, but with a nonzero exit code (rc).  In this case we check the fieldmode value determine set fieldmode=false.
2. Line 304: The command exits abnormally (terminated by exception, or signal, etc); and note there is technically no return code in this case.  In this case we did not get a result so we have to fail securely, meaning fieldmode=true.

So, yes I think the code could be changed to interpret stat as the rc and it would technically still work.  But don't make me get a Linux hacker in here!